### PR TITLE
OSD on line level, recognition by loading script or lang from PAGE

### PR DESCRIPTION
**Pull request by bertsky**
In partial fulfillment of #69 – but I'm afraid osd.traineddata is just too bad for script detection to make this work reliably. Latin vs Cyrillic vs Arabic vs Chinese etc might be easy, but Greek for example does not work on the line level... 

Nevertheless, once we do have script / font detectors on the line level, we could make this work.

- [x] update the PR to the current master
- [x] ~~implement `style_model` (map from `TextStyle` alists to model name)~~
- [x] implement `xpath_model` / `xpath_parameters` (map from PAGE-XML XPath queries to model / variable settings)
- [x] implement `auto_model` (try each loaded model and pick best scoring one)

### CODE REVIEW

File: ocrd_tesserocr/recognize.py [hunk 1]
```diff
@@ -1190,6 +1196,30 @@ def page_update_higher_textequiv_levels(level, pcgts):
                     region_conf /= len(lines)
             region.set_TextEquiv( # replace old, if any
                 [TextEquivType(Unicode=region_unicode, conf=region_conf)])
+    def reinit(self, tessapi, segment):
+        tag = segment.__class__.__name__[:-4]
+        ident = segment.id if hasattr(segment, 'id') else segment.imageFilename
```
Comment (kba) [inline ocrd_tesserocr/recognize.py:1201]: Can be just `segment.id` with https://github.com/OCR-D/core/pull/683

**Review by kba — APPROVED** [review]
Very good, thank you! The mapping can only be as good as `osd` is in detecting script and language, of course, but it's a step towards dynamic model selection and we can use the patterns here in other processors.

**Comment by bertsky** [comment]
> The mapping can only be as good as `osd` is in detecting script and language, of course, 

Now goes beyond that: you can use e.g. `-P script_model '{ "Latn - Latin": "lat+Latin", "Grek - Greek": "grc+Greek" }' -P auto_model true -P model lat+Latin+grc+Greek` to have it try:
1. if `@primaryScript="Latn - Latin"`, then activate `lat+Latin`
2. elif `@primaryScript="Grek - Greek"`, then activate `grc+Greek`
3. otherwise try each loaded model individually and __pick the best-scoring one__ per segment

> but it's a step towards dynamic model selection and we can use the patterns here in other processors.

Yes, but that kind of dynamics will look quite different for each processor implementation (and no other engine has the model variety of Tesseract). I wish we could express any of that on the workflow level...

**Comment by bertsky** [comment]
Note: depends on OCR-D/core#699

**Comment by bertsky** [comment]
@kba, I still had to update our deployment rules for the resmgr changes. CI is now only failing because of the dependency on https://github.com/OCR-D/core/pull/699 – should I wait for your merge (and rebuild of `ocrd/core` on DH), or rather update to core's etree branch here?

**Pull request merged**